### DISCUSSION TITLE
Allow __setitem__ on tracers?

### DISCUSSION THREAD

**Discussion by dionhaefner** [jax-ml/jax#8519] [Ideas]
I am somewhat of a [power](https://github.com/team-ocean/veros) [user](https://github.com/dionhaefner/yahtzotron) [of](https://github.com/dionhaefner/pyhpc-benchmarks) [JAX](https://github.com/mpi4jax/mpi4jax), but I still can't quite get used to JAX's syntax for setitem operations. To be specific, I am talking about this:

```python
# 🥴
arr = arr.at[...].set(0.)

# why not this?!
arr[...] = 0
```

**Concrete problems:**

- It creates friction when porting existing NumPy code to JAX. Every time I tell someone "oh, it's just like NumPy", I mentally add "except this one operation which is used all the time".
- This makes it a lot harder to use both NumPy and JAX as backend for the same code (i.e., conditionally do `import numpy as npx` or `import jax.numpy as npx`).
- A common rookie mistake is to forget `arr = ` on the left hand side, so this silently turns into a no-op.
- It's less concise.
- [People are frequently confused whether this is truly in-place under JIT.](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#array-updates-x-at-idx-set-y)

**Proposal:**

I assume that JAX arrays being immutable is a non-negotiable requirement, but maybe that doesn't need to extend to Tracer objects? In this case, JAX could support `__setitem__` *only* under JIT:

```python
def foo(arr):
    arr[...] = 0.  # Exception
    return arr

@jax.jit
def bar(arr):
    arr[...] = 0.  # works
    return arr
```

What do you think?

**Comment by jakevdp** [comment]
This is an interesting thought, but my first inclination is that we would not want to add this. First of all, one of the guiding design principles behind JAX is to strive for JIT invariance: that is, as much as possible code executed under JIT should behave the same as in op-by-op mode. This would break that assumption.

Additionally, I think it would cause a lot of confusion because we could *only* make it work for `__setitem__`, not for in-place updates like `arr[idx] += 5`. The reason is that `arr[idx] = 5` has its own special method (`__setitem__`), whereas `arr[idx] += 5` must be implemented by allowing `arr[idx]` to return an object that acts as a view into a portion of `arr`, such that when it is modified in-place by `+= 5` the original array is also modified. I'm not sure how we could cleanly implement this using JAX's tracer model (as well as XLA's compilation/computation model!), which is built on a fundamental assumption of immutability.

So the idea of overriding *only* `__setitem__` (and not other in-place operations), and only for tracers (not for arrays), seems like it would miss the goal of simplifying the things users must think about when porting code from numpy to JAX. Rather, it would add another special case of context-dependent operation that users would have to know about and think about when using JAX.

Others may have different reactions though - thanks for starting the discussion!

**Reply by dionhaefner** [reply]
Dang, I knew that I had missed a corner case :) `arr[idx] += 5` is nasty indeed. This is `__getitem__` followed by `__iadd__`, right?

If tracers could become mutable that wouldn't be a problem, but whether that's feasible is of course for you to decide (sounds like it isn't).

**Reply by jakevdp** [reply]
Wait, actually I think I was wrong. `arr[idx] += 5` is a `getitem`, followed by `iadd` on the view, then `setitem`.

If you do `view = arr[idx]; view += 5`, then it's just `getitem` followed by `iadd`.

We may be able to implement the former without mutable tracers in the way you have in mind, but the latter would be pretty difficult.

**Reply by dionhaefner** [reply]
Maybe I'm a bit dense here, but how would you even do `__setitem__` without mutable tracers? It's not like you can return a new object - what would that be assigned to?

**Reply by jakevdp** [reply]
I think you could fake it in the same way we currently fake `__iadd__` on tracers (there have been discussions about explicitly disabling in-place arithmetic operators on JAX objects, but nothing has come of it yet). So long as you don't need mutable view semantics, I think it can be done without breaking too many existing assumptions.

Edit: actually, I see your point about not having any object to assign the update to. It may be hard...
